### PR TITLE
Add Timed Walk Test & Improve Rendering Possibilities for the QuestionnaireView

### PR DESCRIPTION
**Pull request by dguo8412**
# Add Timed Walk Test & Improve Rendering Possibilities for the `QuestionnaireView`

## :gear: Release Notes 
- Adds the `TimedWalkTestView`
- Adds the cancellation configuration to the `QuestionnaireView` 

## :books: Documentation
- Adds documentation for all code changes.

## :white_check_mark: Testing
- Adds automated tests for the timed walk test.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift [hunk 1]
```diff
@@ -35,10 +35,10 @@ struct ORKOrderedTaskView: UIViewControllerRepresentable {
                 
                 switch reason {
                 case .completed:
-                    let fhirResponse = taskViewController.result.fhirResponse
-                    fhirResponse.subject = Reference(reference: FHIRPrimitive(FHIRString("My Patient")))
+                    let response = taskViewController.result
                     
-                    await questionnaireResponse(fhirResponse)
+                    // what is this doing down here?
+                    await timedWalkResponse(response)
```
Comment (dguo8412) [inline Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift:41]: I am unsure if I'm doing this correctly here. Is it now up to the developer using our Spezi framework to do what they want with the response?

File: Sources/SpeziQuestionnaire/TimedWalkView.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    
+    
+    /// Creates a ResearchKit navigable task from a questionnaire
+    /// - Parameter questionnaire: a questionnaire
+    /// - Returns: a ResearchKit ordered task
+    private func createTask() -> ORKOrderedTask? {
+        // Create a navigable task from the Questionnaire
+        do {
+            return try timedWalk(withIdentifier: identifier, intendedUseDescription: "", distanceInMeters: distanceInMeters, timeLimit: timeLimit, turnAroundTimeLimit: turnAroundTimeLimit, includeAssistiveDeviceForm: false, options: ORKPredefinedTaskOption(rawValue: 0))
```
Comment (dguo8412) [inline Sources/SpeziQuestionnaire/TimedWalkView.swift:76]: This is the part this is most confusing to me. I am calling this ResearchKit method: https://researchkit.org/docs/Classes/ORKOrderedTask.html#//api/name/timedWalkTaskWithIdentifier:intendedUseDescription:distanceInMeters:timeLimit:turnAroundTimeLimit:includeAssistiveDeviceForm:options:

At first, Xcode autofilled the method for me, but now it doesn't show up and it "cannot be found in scope" even though I import ResearchKit at the top.

File: Sources/SpeziQuestionnaire/TimedWalkView.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        self.turnAroundTimeLimit = turnAroundTimeLimit
+        self._isPresented = isPresented
+        self.timedWalkResponse = timedWalkResponse
+    }
+    
+    
+    /// Creates a ResearchKit navigable task from a questionnaire
+    /// - Parameter questionnaire: a questionnaire
+    /// - Returns: a ResearchKit ordered task
+    private func createTask() -> ORKOrderedTask? {
```
Comment (dguo8412) [inline Sources/SpeziQuestionnaire/TimedWalkView.swift:73]: This function has questionnaire as an input for the questionnaireView. What should I have as an input here? Do I even need an input?

File: Sources/SpeziQuestionnaire/TimedWalkView.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+        
+    public var body: some View {
+        if let task = createTask() {
+            ORKOrderedTaskView(
+                tasks: task,
+                isPresented: $isPresented,
+                timedWalkResponse: { response in
+                    await timedWalkResponse?(response)
+                    await questionnaireDataSource.add(response)
```
Comment (dguo8412) [inline Sources/SpeziQuestionnaire/TimedWalkView.swift:36]: What is the datasource for? To store responses? How many responses are returned by timedWalk? I thought it was only 1 response.

File: Sources/SpeziQuestionnaire/TimedWalkView.swift [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                },
+                tintColor: .accentColor
+            )
+                .ignoresSafeArea(.container, edges: .bottom)
+                .ignoresSafeArea(.keyboard, edges: .bottom)
+        } else {
+            Text("QUESTIONNAIRE_LOADING_ERROR_MESSAGE")
```
Comment (dguo8412) [inline Sources/SpeziQuestionnaire/TimedWalkView.swift:43]: Where was this string stored again? I want to change the error message and variable name

File: Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift (hunk 1 above)
Comment (dguo8412) [reply Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift:41]: Also, this is the incorrect type with error message "cannot convert value of type 'ORKTaskResult' to expected argument type 'ORKTimedWalkResult'".

I also tried casting it in line 38 above, but they are incompatible. I might need a little refresher on class inheritance (both ORKTaskResult and ORKTimedWalkResult are subclasses of ORKResult, but I am unsure how to syntactically convert them)

File: Sources/SpeziQuestionnaire/TimedWalkView.swift (hunk 3 above)
Comment (PSchmiedmayer) [reply Sources/SpeziQuestionnaire/TimedWalkView.swift:73]: @dguo8412 Discussed this today in our meeting, let me know if you have any follow up questions 👍

File: Sources/SpeziQuestionnaire/TimedWalkView.swift (hunk 5 above)
Comment (PSchmiedmayer) [reply Sources/SpeziQuestionnaire/TimedWalkView.swift:43]: Good catch! This one is actually not localized and an oversight! You would need to create a strings catalog in a Resources folder of the Swift Package target. You can learn more about this in the following WWDC session at the very end (around minute 30): https://developer.apple.com/wwdc23/10155

File: Sources/SpeziQuestionnaire/TimedWalkView.swift (hunk 3 above)
Comment (dguo8412) [reply Sources/SpeziQuestionnaire/TimedWalkView.swift:73]: I don't remember if we discussed this one in our meeting. Do we need an input for creating a task here? It seems to work fine without one.

File: Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift [hunk 6]
```diff
@@ -35,10 +35,12 @@ struct ORKOrderedTaskView: UIViewControllerRepresentable {
                 
                 switch reason {
                 case .completed:
-                    let fhirResponse = taskViewController.result.fhirResponse
-                    fhirResponse.subject = Reference(reference: FHIRPrimitive(FHIRString("My Patient")))
-                    
-                    await questionnaireResponse(fhirResponse)
+                    // add a for loop to check for the ORKTimedWalkResult
```
Comment (dguo8412) [inline Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift:38]: I had some syntactical trouble implementing this for loop. Any tips on resolving this?

Screenshot 2023-10-04 at 1 17 23 AM

Comment (PSchmiedmayer) [reply Sources/SpeziQuestionnaire/ORKOrderedTaskView.swift:38]: You will probably have to add a guard statement before that first unwraps the optional value into e.g. a `results` constant and then use that in the loop 👍

File: Sources/SpeziQuestionnaire/TimedWalkView.swift (hunk 3 above)
Comment (PSchmiedmayer) [reply Sources/SpeziQuestionnaire/TimedWalkView.swift:73]: You are correct. You currently create the timed walk task right in the function. In the long run, you might have a `createActiveTask` function that takes the input and then creates a proper `ORK...` type but it is fine if this happens in the current function signature for now.

File: Sources/SpeziQuestionnaire/TimedWalkView.swift [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+    
+    
+    /// - Parameters:
+    ///   - questionnaire: The  `Questionnaire` that should be displayed.
+    ///   - isPresented: Indication from the questionnaire view if should be presented (not "Done" pressed or cancelled).
+    ///   - completionStepMessage: Optional completion message that can be appended at the end of the questionnaire.
+    ///   - questionnaireResponse: Optional response closure that can be used to manually obtain the `QuestionnaireResponse`.
+    public init(
+        identifier: String,
```
Comment (PSchmiedmayer) [inline Sources/SpeziQuestionnaire/TimedWalkView.swift:54]: I would suggest that we provide a reasonable default here that just generates a unique string for every time that the task is generated:
```suggestion
        identifier: String = UUID().uuidString,
```

File: Sources/SpeziQuestionnaire/TimedWalkView.swift [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        // Create a navigable task from the Questionnaire
+        return ORKOrderedTask.timedWalk(
+            withIdentifier: identifier,
+            intendedUseDescription: "",
+            distanceInMeters: distanceInMeters,
+            timeLimit: timeLimit,
+            turnAroundTimeLimit: turnAroundTimeLimit,
+            includeAssistiveDeviceForm: false,
+            options: ORKPredefinedTaskOption(rawValue: 0)
+        )
```
Comment (PSchmiedmayer) [inline Sources/SpeziQuestionnaire/TimedWalkView.swift:83]: I have taken a closer look at the RK documentation and it seems like [fitnessCheckTaskWithIdentifier:intendedUseDescription:walkDuration:restDuration:options:](https://researchkit.org/docs/Classes/ORKOrderedTask.html#//api/name/fitnessCheckTaskWithIdentifier:intendedUseDescription:walkDuration:restDuration:options:) seems to be the better task suited for the 6 minute walk test. I tried the application and it ask you to walk a certain distance using the timed walk. We are looking for a way to walk 6 minutes and get the data from there.

File: Tests/UITests/UITests.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -388,7 +388,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				CURRENT_PROJECT_VERSION = 1;
 				DEVELOPMENT_ASSET_PATHS = "";
-				DEVELOPMENT_TEAM = 64FJ2MWNP4;
+				DEVELOPMENT_TEAM = 4DR2J9J8WP;
```
Comment (PSchmiedmayer) [inline Tests/UITests/UITests.xcodeproj/project.pbxproj:391]: Please ensure that you reset changes like this before we merge the PR.

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
Thank you for the first draft! I added some comments and suggestions on how to approach the generalization in individual comments as the review 👍

File: Sources/SpeziQuestionnaire/TimedWalkView.swift (hunk 8 above)
Comment (dguo8412) [reply Sources/SpeziQuestionnaire/TimedWalkView.swift:83]: This does look like a much more appropriate task

**Comment by PSchmiedmayer** [comment]
@dguo8412 Thank you for all the work here! Feel free to tag me for an other review once you have finished the UI tests and addressed all other red warnings. Great to see the progress in this PR and to have you in the office to work on it! 👍

File: Sources/SpeziTimedWalkTest/TimedWalkTestView.swift [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            }
+            .navigationTitle("Timed Walk Test")
+            .modifier(TimedWalkTestCancelModifier())
+            .environment(walkTestViewModel)
+    }
+
+    
+    public init(
+        timedWalkTest: TimedWalkTest = TimedWalkTest(),
+        completion: @escaping (Result<TimedWalkTestResult, TimedWalkTestError>) -> Void
```
Comment (PSchmiedmayer) [inline Sources/SpeziTimedWalkTest/TimedWalkTestView.swift:54]: We should replace this with something beyond the `Result` type to model the cancel type as we do in the questionnaire target.

Comment (dguo8412) [reply Sources/SpeziTimedWalkTest/TimedWalkTestView.swift:54]: I'll take a look!

**Pull request merged**